### ISSUE TITLE
Resolving names in VSC

### ISSUE THREAD

**Issue by lefebvresam** [mbed-ce/mbed-os#243]
In Visual Studio Code there is a name resolver running in the background. But because of the fragmentation of the mbed code and the generation of some code parts during the build, the name resolver is not able to resolve everything. This can sometime be annoying if your are dealing with objects like Thread.

![image](https://github.com/mbed-ce/mbed-os/assets/15315428/219c51b2-9144-4145-881d-f6f48e53dd38)

```
#include "Thread.h"
using namespace rtos;

    private members:
        Thread _thread;
        Thread* _buzzerthread;
```

Then Thread is no longer underlined in red, but is pointing to the test classes, not the real ones. And they are different.
If you point to the real one */rtos/inlcude/rtos/Thread.h* the class is not found because of undifined macro's.

If you initialize and want tor recognize the os proirities:

I tested by printing out osPriorityBelowNormal and it is 16, so the real header used is:
cmsis/CMSIS_5/CMSIS/RTOS2/Include/cmsis_os2.h

```
#include "cmsis/CMSIS_5/CMSIS/RTOS2/Include/cmsis_os2.h"

In constructor
_thread(osPriorityBelowNormal, 1024, nullptr, "display")
```

The test class is pointing to another (limited) version of cmsis_os2.h

This is very confusing and I don't know if it should be interesting to have a topic about SVC and name resolving, because it's an essential feature in programming. I don't know if it is technically possible to resolve to the right header names by VSC.

**Comment by JojoS62** [comment]
do you have the setting "configurationProvider": "ms-vscode.cmake-tools" in c_cpp_properties.json?
And also the cmake-tools extension installed of course. Then the include path are used that the compiler generates and Intellisense works much better than with other settings.

**Comment by lefebvresam** [comment]
Do you mean this. Has no effect.
![image](https://github.com/mbed-ce/mbed-os/assets/15315428/984dad8b-1c35-497a-8168-70d42ef35434)

**Comment by JojoS62** [comment]
Yes, and do have the CMake Extension installed?

**Comment by lefebvresam** [comment]
![image](https://github.com/mbed-ce/mbed-os/assets/15315428/b4b3f8cc-af2d-4f5d-8780-3671524a4002)

**Comment by JojoS62** [comment]
that looks fine. I have also this setting in c_cpp_properties.json:
"intelliSenseMode": "${default}",

**Comment by lefebvresam** [comment]
I don't find one of these files in my test project. Where are they located and when are they created?

**Comment by JojoS62** [comment]
compile_commands.json is in my ./build/target folder and the export of this file the default setting in the cmake-extension.
To get the subidr target created, check step 6 in:
https://github.com/mbed-ce/mbed-os/wiki/Project-Setup:-VS-Code

The c_cpp_prorperties.json is located in the .vscode directory.

**Comment by lefebvresam** [comment]
I have non of these files even after I generated the build folder again with the line you mentioned in settings.json:

![image](https://github.com/mbed-ce/mbed-os/assets/15315428/ccd8a895-dd45-4205-8b7c-9d824fac9731)

**Comment by JojoS62** [comment]
strange, I have checked again, the compile_commands.json is created when I select the configuration in the CMake extension and the Mbed configuration has finished.
Also the cmake.buildDirectory is not created as in the setting, it does not have the target name. So it looks like the extension is not running properly or is missing something.

![grafik](https://github.com/mbed-ce/mbed-os/assets/6775562/0bfbe806-7c7f-454c-a04c-1304e175f606)

**Comment by lefebvresam** [comment]
https://github.com/microsoft/vscode/issues/129466 looks more like a bug

**Comment by JojoS62** [comment]
I found this issue also, but it is old. Usually things get fixed soon, so in this case it may happen under some special circumstances. 

![grafik](https://github.com/mbed-ce/mbed-os/assets/6775562/98e20e69-3f04-473e-8fb6-fd32a8b9ee68)

When this CMake View is selected, you see the configuration. When the configuration is changed, it runs the cmake configure with the menu from cmake-variants.yaml.

**Comment by lefebvresam** [comment]
![image](https://github.com/mbed-ce/mbed-os/assets/15315428/37f4f2e1-03cd-48ae-bbef-05a36778697c)

**Comment by lefebvresam** [comment]
I did the while procedure and after run and debug nothing happens:

![image](https://github.com/mbed-ce/mbed-os/assets/15315428/ee63e3ba-8ed8-4ae0-8a91-5633902de200)

But I created a build shell script for it instead of using the baked in functionality for cmake.
I was also wondering why compiling in home folder should be faster in case of WSL? My experiënce is that the best solution is to connect to a fast Linux build laptop next to you on the table. Then you unload your desk pc during compilation.

**Comment by JohnK1987** [comment]
Project was probably not created according to MbedCE instructions, that is the reason why there was different behaviour.

**Comment by lefebvresam** [comment]
In my case there are two ways to compile and program:

1/ The standard way F7, using cmake. In that case I have a build folder with a compile_commands.json file being generated automatically.

2/ Using my own build script _'sh build.sh -f'_  (-f to flash also). In that case the compile_commands.json is not generated.
See [build.sh.txt](https://github.com/user-attachments/files/16312742/build.sh.txt)

I have currently no additional files in  .vscode/settings.json. No settings.json and no c_cpp_properties.json. This is however not practical because this file will not be checked in with source control and this is lost after each installation.

I think the setting c_cpp_properties.json is overwritten in:
![image](https://github.com/user-attachments/assets/6c8d253f-fab1-4062-bee4-75eeac98780e)
And this survives re-installs of development pc because it's probably a local setting that remains persistent. 

In both cases I do not longer face issues with intellisense, but this is maybe due to the fact that everything is already cached locally and I encountered this issue the first time I used my shell script without compiling with cmake. If you compile with cmake the first time maybe from that point everything is cached.
